src/Altinn.App.Core/Features/Validation/ValidationService.cs

```diff
@@ -0,0 +1,156 @@
+
+        var dataElementsIssuesTask = Task.WhenAll(validators.Select(async v =>
+        {
+            try
+            {
+                _logger.LogDebug("Start running validator {validatorName} on {dataType} for data element {dataElementId} in instance {instanceId}", v.GetType().Name, dataElement.DataType, dataElement.Id, instance.Id);
+                var issues = await v.ValidateDataElement(instance, dataElement, dataType);
+                issues.ForEach(i => i.Source = v.ValidationSource); // Ensure that the source is set to the validator source
+                return issues;
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Error while running validator {validatorName} on {dataType} for data element {dataElementId} in instance {instanceId}", v.GetType().Name, dataElement.DataType, dataElement.Id, instance.Id);
+                throw;
+            }
+        }));
+
+        // Run extra validation on form data elements with app logic
+        if(dataType.AppLogic?.ClassRef is not null)
+        {
+            Type modelType = _appModel.GetModelType(dataType.AppLogic.ClassRef);
+
+            Guid instanceGuid = Guid.Parse(instance.Id.Split("/")[1]);
+            string app = instance.AppId.Split("/")[1];
+            int instanceOwnerPartyId = int.Parse(instance.InstanceOwner.PartyId);
+            var data = await _dataClient.GetFormData(instanceGuid, modelType, instance.Org, app, instanceOwnerPartyId, Guid.Parse(dataElement.Id)); // TODO: Add method that accepts instance and dataElement
+            var formDataIssuesDictionary = await ValidateFormData(instance, dataElement, dataType, data, null);
+
+            return (await dataElementsIssuesTask).SelectMany(x=>x)
+                .Concat(formDataIssuesDictionary.SelectMany(kv=>kv.Value))
+                .ToList();
+        }
+
+        return (await dataElementsIssuesTask).SelectMany(x=>x).ToList();
+    }
+
+    /// <inheritdoc/>
+    public async Task<Dictionary<string, List<ValidationIssue>>> ValidateFormData(Instance instance, DataElement dataElement, DataType dataType, object data,
+        List<string>? changedFields = null, List<string>? ignoredValidators = null)
+    {
+        ArgumentNullException.ThrowIfNull(instance);
+        ArgumentNullException.ThrowIfNull(dataElement);
+        ArgumentNullException.ThrowIfNull(dataElement.DataType);
+        ArgumentNullException.ThrowIfNull(data);
+
+        // Locate the relevant data validator services from normal and keyed services
+        var dataValidators = _serviceProvider.GetServices<IFormDataValidator>()
+            .Where(dv => dv.DataType == "*" || dv.DataType == dataType.Id)
+            // .Concat(_serviceProvider.GetKeyedServices<IFormDataValidator>(dataElement.DataType))
+            .Where(dv => ignoredValidators?.Contains(dv.ValidationSource) != true)
+            .Where(dv => changedFields is null ? true : dv.ShouldRun(changedFields))
```

Comment:
As suggested by codeQL
```suggestion
            .Where(dv => changedFields is null || dv.ShouldRun(changedFields))
```